src/csharp/Microsoft.Spark/ML/Feature/Pipeline.cs

```diff
@@ -0,0 +1,121 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System;
+using System.Linq;
+using System.Reflection;
+using Microsoft.Spark.Interop;
+using Microsoft.Spark.Interop.Ipc;
+using Microsoft.Spark.Sql;
+using Microsoft.Spark.Interop.Internal.Java.Util;
+
+namespace Microsoft.Spark.ML.Feature
+{
+
+    /// <summary>
+    /// <see cref="Pipeline"/> A simple pipeline, which acts as an estimator. 
+    /// A Pipeline consists of a sequence of stages, each of which is either an Estimator or a Transformer.
+    /// When Pipeline.fit is called, the stages are executed in order. If a stage is an Estimator, its 
+    /// Estimator.fit method will be called on the input dataset to fit a model. Then the model, which is a 
+    /// transformer, will be used to transform the dataset as the input to the next stage. 
+    /// If a stage is a Transformer, its Transformer.transform method will be called to produce the 
+    /// dataset for the next stage. The fitted model from a Pipeline is a PipelineModel, which consists of 
+    /// fitted models and transformers, corresponding to the pipeline
+    /// stages. If there are no stages, the pipeline acts as an identity transformer.
+    /// </summary>
+    public class Pipeline : ScalaEstimator<PipelineModel>, ScalaMLWritable, ScalaMLReadable<Pipeline>
+    {
+        private static readonly string s_pipelineClassName = "org.apache.spark.ml.Pipeline";
+
+        /// <summary>
+        /// Creates a <see cref="Pipeline"/> without any parameters.
+        /// </summary>
+        public Pipeline() : base(s_pipelineClassName)
+        {
+        }
+
+        /// <summary>
+        /// Creates a <see cref="Pipeline"/> with a UID that is used to give the
+        /// <see cref="Pipeline"/> a unique ID.
+        /// </summary>
+        /// <param name="uid">An immutable unique ID for the object and its derivatives.</param>
+        public Pipeline(string uid) : base(s_pipelineClassName, uid)
+        {
+        }
+
+        internal Pipeline(JvmObjectReference jvmObject) : base(jvmObject)
+        {
+        }
+
+        /// <summary>
+        /// Set the stages of pipeline instance.
+        /// </summary>
+        /// <param name="value">
+        /// A sequence of stages, each of which is either an Estimator or a Transformer.
+        /// </param>
+        /// <returns><see cref="Pipeline"/> object</returns>
+        public Pipeline SetStages(ScalaPipelineStage[] value) =>
+            WrapAsPipeline((JvmObjectReference)SparkEnvironment.JvmBridge.CallStaticJavaMethod(
+                "org.apache.spark.api.dotnet.DotnetHelper", "setPipelineStages",
```

Comment:
Curious if this was necessary.  Did calling the following not work ?

```cs
Reference.Invoke("setStages", value)
```
```

Reply:
It doesn't work because you can't directly set a java.util.ArrayList to the scala's method which only accepts scala array. That's also why I added this helper function in the scala side:
def setPipelineStages(pipeline: Pipeline, value: java.util.ArrayList[_ <: PipelineStage]): Pipeline =
    pipeline.setStages(value.asScala.toArray)